Document snowflake account setup (#236)
cagov/data-infrastructure: docs/new-project-setup.md

# New Project Setup

The DSE team regularly creates new Snowflake accounts in our Snowflake org.
We do this instead of putting all of our data projects into our main account for a few reasons:

1. At the end of a project, we often want to transfer account ownership to our partners.
    Having it separated from the start helps that process.
1. We frequently want to add our project champion or IT partners to our account as admins.
    This is safer if project accounts are separate.
1. We often want to have accounts in a specific cloud and region for compliance or data transfer regions.
1. Different projects may require different approaches to account-level operations like OAuth/SAML.

Here we document the steps to creating a new Snowflake account from scratch.

## Prerequisites

### Obtain permissions in Snowflake

In order to create a new account, you will need access to the `orgadmin` role.
If you have `accountadmin` in the primary Snowflake account, you can grant it to yourself:

```sql
USE ROLE accountadmin;

> GRANT ROLE orgadmin TO USER <YOUR-USER>;

Reviewer: I think it would be a good idea to add code here for how to revoke access to the `orgadmin` role as well.

Author: Done

One example of such a tool is (as of this writing) the Snowflake Terraform Provider.

The recommended workaround for this is to add a key pair to your account for use with those tools.

1. Follow the instructions given [here](https://docs.snowflake.com/en/user-guide/key-pair-auth#configuring-key-pair-authentication)
    to generate a key pair and add the public key to your account.
    Keep the key pair in a secure place on your device.
    [This gist](https://gist.github.com/ian-r-rose/1c714ee04be53f7a3fd80322e1a22c27)
    can be helpful for quickly creating a new encrypted key pair.
    Usage of the script looks like:
    ```bash
    bash generate_encrypted_key.sh <key-name> <passphrase>
    ```
1. In your local `.bash_profile` or an `.env` file, add environment variables for
    `SNOWFLAKE_PRIVATE_KEY_PATH` and (if applicable) `SNOWFLAKE_PRIVATE_KEY_PASSPHRASE`.

### Apply a session policy

By default, Snowflake logs out user sessions after four hours of inactivity.
ODI's information security policies prefer that we log out after one hour of inactivity for most accounts,
and after fifteen minutes of inactivity for particularly sensitive accounts.

> !!! note

Reviewer: !!! note is not working in this file or any of our others anymore, not sure why!

Author: This uses a markdown extension that is enabled in our docs build, but does not render in GitHub

Ask a DSE AWS admin to give you access to these buckets,
and [configure your AWS credentials](./setup.md#aws).

### Install terraform dependencies

You can install Terraform using whatever approach makes sense for your system,
including using `brew` or `conda`.

Here is a sample for installing the dependencies using `conda`:

```bash
conda create -n infra python=3.10  # create an environment named 'infra'
conda activate infra  # activate the environment
conda install -c conda-forge terraform tflint  # Install terraform and tflint
```

## Snowflake account setup

### Create the account

1. Assume the `ORGADMIN` role
1. Under the "Admin" side panel, go to "Accounts" and click the "+ Account" button:
    1. Select the cloud and region appropriate to the project. The region should be in the United States.
    1. Select "Business Critical" for the Snowflake Edition.
    1. You will be prompted to create an initial user with `ACCOUNTADMIN` privileges. This should be you.
        You will be prompted to create a password for your user. Create one using your password manager,
        but know that it will ask you change your password upon first log-in.

>     1. Save the Account Locator and URL for your new account.

Reviewer: ```suggestion
    1. Save the Account Locator and Account Locator URL for your new account.
```
Account URL and Account Locator URL are different so I am suggesting this additional language for more clarity

1. Log into your new account. You should be prompted to change your password. Save the updated password in your password manager.

### Enable multi-factor authentication for your user

1. Ensure the Duo Mobile app is installed on your phone.

> 1. In the upper-left corner, click on your user, and select "Profile"

Reviewer: ```suggestion
1. In the upper-left corner of Snowflake, click on your username, and select "Profile"
```

1. At the bottom of the dialog, select "Enroll" to enable multi-factor authentication.
1. Follow the instructions to link the new account with your Duo app.

### Set up key pair authentication

Certain Snowflake clients don't properly cache MFA tokens,
which means that using them can generate dozens or hundreds of MFA requests on your phone.
At best this makes the tools unusable, and at worst it can lock your Snowflake account.
One example of such a tool is (as of this writing) the Snowflake Terraform Provider.

The recommended workaround for this is to add a key pair to your account for use with those tools.

1. Follow the instructions given [here](https://docs.snowflake.com/en/user-guide/key-pair-auth#configuring-key-pair-authentication)

>     to generate a key pair and add the public key to your account.

Reviewer: ```suggestion
    to generate a key pair and add the public key to your account.
    
    Use `pbcopy < _your_public_key_file_name_.pub` to copy the contents of your public key. Be sure to remove the commented lines before and after.
```

    Keep the key pair in a secure place on your device.

>     [This gist](https://gist.github.com/ian-r-rose/1c714ee04be53f7a3fd80322e1a22c27)

Reviewer: I did not use this gist or the bash code below. if they are optional I think separating them out and denoting that would be helpful

Author: The gist is just a copy/paste of the bash commands in the linked docs. It's quite helpful when you are creating a lot of key pairs, which is the case for setting up an account. It is optional (which is what I meant to indicate by "can be helpful"), I can try to state it a bit more strongly.

    but at the time of this writing the Snowflake Terraform provider does not support session policies.

After the Snowflake account is created, run the following script in a worksheet
to set the appropriate session policy:

```sql
use role sysadmin;
create database if not exists policies;
create session policy if not exists policies.public.account_session_policy
  session_idle_timeout_mins = 60
  session_ui_idle_timeout_mins = 60
;
use role accountadmin;
-- alter account unset session policy;  -- unset any previously existing session policy
alter account set session policy policies.public.account_session_policy;
```

### Add IT-Ops representatives

TODO: establish and document processes here.

### Set up Okta SSO and SCIM

TODO: establish and document processes here.

## Create project git repository

Create a new git repository from the CalData Infrastructure Template
following the instructions [here](https://github.com/cagov/caldata-infrastructure-template#usage).

Once you have created the repository, push it to a remote repository in GitHub.
There are some GitHub actions that will fail because the repository is not yet

> configured to work with the new Snowflake account.

Reviewer: I may not be able to go farther than this because the linked instructions are insufficient and I am still trying to figure out how to push the local repository to a non-existent remote repo.

Reviewer: There are a lot of steps I had to take that were not listed in those docs that I'll have to add later.

Reviewer: A quick tl;dr is:

- I had to create a local folder first to house all the files generated by the copier command
- That said i did not need to `cd` to that repo bc i was already in it after creating it
- I had to create a remote repo
- I had to run commands to link my local repo to that remote one
- I now have to create a PAT and tick something about workflows

Reviewer: Got past this, will update the docs with the necessary info later

- links i used for reference:
   - https://github.com/orgs/community/discussions/26254
   - https://stackoverflow.com/questions/30068298/git-fatal-could-not-read-from-remote-repository-please-make-sure-you-have-th
   - https://docs.github.com/en/migrations/importing-source-code/using-the-command-line-to-import-source-code/adding-locally-hosted-code-to-github

Author: > *  I had to create a local folder first to house all the files generated by the copier command

This is probably an unintended consequence of https://github.com/cagov/caldata-infrastructure-template/pull/21 . We should update the docs over there to reflect that you need to create a directory first now.

> *  That said i did not need to cd to that repo bc i was already in it after creating it

I don't understand this. After creating a directory, you are typically not placed in it, i.e., 
```bash
mkdir my-project-repo
```
does not place you in `my-project-repo`.

> * I had to create a remote repo
> * I had to run commands to link my local repo to that remote one

I'm currently thinking that instructions for how to create a GitHub repo are out-of-scope for these docs, i.e., this shouldn't cover every sub-step that needs to be done, but instead the steps that are specific to this project-set-up process.

> * I now have to create a PAT and tick something about workflows

I don't understand what you are referring here, but creating and pushing to a GitHub repo does not require a personal access token, so something seems to have gone awry.

We will create two separate deployments of the project infrastructure,
one for development, and one for production.
In some places we will refer to project name and owner as `<project>` and `<owner>`, respectively,
following our [naming conventions](./naming-conventions.md).
You should substitute the appropriate names there.

### Create the dev configuration

1. In the new git repository, create a directory to hold the development Terraform configuration:
    ```bash
    mkdir -p terraform/environments/dev/
    ```
    The location of this directory is by convention, and subject to change.
1. Copy the terraform configuration from
    [here](https://github.com/cagov/data-infrastructure/blob/main/terraform/snowflake/environments/dev/main.tf)
    to your `dev` directory.
1. In the "elt" module of `main.tf`, change the `source` parameter to point to
    `"github.com/cagov/data-infrastructure.git//terraform/snowflake/modules/elt?ref=<ref>"`
    where `<ref>` is the short hash of the most recent commit in the `data-infrastructure` repository.
1. In the `dev` directory, create a new backend configuration file called `<owner>-<project>-dev.tfbackend`.

>     The file will point to the S3 bucket in which we are storing terraform state:

Reviewer: are we meant to copy this code in the newly created file? if so i think we should say that explicitly.
```suggestion
    The file will point to the S3 bucket in which we are storing terraform state. Copy the following code into the file you just created:
```

    ```hcl
    bucket = "dse-snowflake-dev-terraform-state"
    dynamodb_table = "dse-snowflake-dev-terraform-state-lock"
    key = "<owner>-<project>-dev.tfstate"
    region = "us-west-2"
    ```

> 1. In the `dev` directory, create a terraform variables file called `terraform.tfvars`,

Reviewer: are we meant to populate the new file with just the two lines of code listed below? or is this info about "elt" module variables that should also be there? @ian-r-rose

    and populate the "elt" module variables. These variables may expand in the future,
    but at the moment they are just the new Snowflake account locator and the environment
    (in this case `"DEV"`):
    ```hcl
    locator     = "<account-locator>"
    environment = "DEV"
    ```
1. Initialize the configuration:
    ```bash
    terraform init -backend-config <owner>-<project>-dev.tfbackend
    ```
1. Include both Mac and Linux provider binaries in your terraform lock file.
    This helps mitigate differences between CI environments and ODI Macs:
    ```bash
    terraform providers lock -platform=linux_amd64 -platform=darwin_amd64
    ```
1. Add your new `main.tf`, `terraform.tfvars`, `<owner>-<project>-dev.tfbackend`,
    and terraform lock file to the git repository. Do not add the `.terraform/` directory.

### Deploy the dev configuration

1. Ensure that your local environment has environment variables set for `SNOWFLAKE_ACCOUNT`,

>     `SNOWFLAKE_USER`, `SNOWFLAKE_PRIVATE_KEY_PATH`,  and `SNOWFLAKE_PRIVATE_KEY_PASSPHRASE`,

Reviewer: unclear where to get the information for the last two variables and sample values are not provided like they are [here](https://github.com/cagov/data-infrastructure/blob/c751c55c6fa634a8860686fb7838817bbc9a889b/docs/setup.md)

Author: These are the private key and password you created in the above steps for setting up key pair authentication

    and that they are set to your new account, rather than any other accounts.

> 1. Run `terraform plan` to see the plan for the resources that will be created.

Reviewer: plan failed and it is unclear what to do to fix it. are there any basic troubleshooting tips that make sense to be added here?

Author: I don't think that troubleshooting steps are in-scope here. This is intended to be project set-up docs for our team for reference when setting up a new project.

Make sure to provide enough information to disambiguate the key pair from others stored in the vault,
including:

* The account locator
* The account name (distinct from the account locator)
* The service account name
* The public key
* The private key

### Set up GitHub actions secrets

You need to configure secrets in GitHub actions
in order for the service accounts to be able to connect to your Snowflake account.
From the repository page, go to "Settings", then to "Secrets and variables", then to "Actions".

Add the following repository secrets:

| Variable | Value |
|----------|-------|
| `SNOWFLAKE_ACCOUNT` | new account locator |
| `SNOWFLAKE_USER_DEV` | `GITHUB_ACTIONS_SVC_USER_DEV` |
| `SNOWFLAKE_USER_PRD` | `GITHUB_ACTIONS_SVC_USER_PRD` |
| `SNOWFLAKE_PRIVATE_KEY_DEV` | dev service account private key |
| `SNOWFLAKE_PRIVATE_KEY_PRD` | prd service account private key |

### Enable GitHub pages for the repository

The repository must have GitHub pages enabled in order for it to deploy and be viewable.

1. From the repository page, go to "Settings", then to "Pages".
1. Under "GitHub Pages visibility" select "Private" (unless the project is public!).

> 1. Under "Build and deployment" select "Deploy from a branch" and choose "gh-pages" as your branch.

Reviewer: should there be a "gh-pages" branch already? bc there isn't one for me @ian-r-rose

Author: It should be there after the first time the docs action successfully completes

Reviewer: My GH actions fail due to an inability to install logbook. >> "Cannot install logbook." Is that expected?

Author: do you have a link?

Reviewer: https://github.com/cagov/britt_test/actions/runs/6775785678/job/18418081852#step:7:122 @ian-r-rose

Reviewer: added you to the repo @ian-r-rose

Author: Try committing your `poetry.lock` file to the repo

Reviewer: I don't have that file. I am not seeing it mentioned in the set up docs either.

Author: We can add it here, there's some discussion of poetry in the local setup docs but not here.

I'm also not certain it will fix the issue you are seeing. Try running `poetry install`, that should generate the file

Reviewer: new errors: https://github.com/cagov/britt_test/actions/runs/6776890533/job/18419372049#step:8:30
and: https://github.com/cagov/britt_test/actions/runs/6776890551 @ian-r-rose

Author: Looks like the credentials in the repo secrets are not right (that's dbt complaining that it can't connect to Snowflake)

Reviewer: are we the value for `SNOWFLAKE_USER_DEV` to literally be `GITHUB_ACTIONS_SVC_USER_DEV` and the same for prd? @ian-r-rose
